src/core/algorithms/gfd/gfd_miner/gfd_miner.cpp

```diff
@@ -445,87 +441,212 @@ void GfdMiner::AddCompacted(std::vector<SingleRule>& rules, model::graph_t& patt
+
+        if (attrs_info.find(fst_label) == attrs_info.end() ||
+            attrs_info.find(snd_label) == attrs_info.end()) {
             continue;
         }
+        for (auto const& fst_name : attrs_info.at(fst_label)) {
+            model::Gfd::Token fst_token{fst, fst_name.first};
+            for (auto const& snd_name : attrs_info.at(snd_label)) {
+                std::set<std::string> intersect;
+                std::set_intersection(fst_name.second.begin(), fst_name.second.end(),
+                                      snd_name.second.begin(), snd_name.second.end(),
+                                      std::inserter(intersect, intersect.begin()));
+                if (intersect.empty()) {
+                    continue;
+                }
+
+                model::Gfd::Token snd_token{snd, snd_name.first};
 
-        model::Gfd gfd = {pattern, std::vector<model::Gfd::Literal>{},
-                          std::vector<model::Gfd::Literal>{l}};
-        if (Validate(gfd, embeddings)) {
-            rules.emplace_back(std::vector<model::Gfd::Literal>{}, l);
+                model::Gfd::Literal l = {fst_token, snd_token};
+                std::set<std::size_t> satisfied_embeddings;
+                std::size_t i = 0;
+                for (Embedding const& embedding : embeddings) {
+                    if (LiteralSatisfied(l, embedding, pattern)) {
+                        satisfied_embeddings.insert(i);
+                    }
+                    i++;
+                }
+                if (satisfied_embeddings.size() >= sigma_) {
+                    satisfied_embeddings_set.try_emplace(l, satisfied_embeddings);
+                    result.emplace_back(fst_token, std::move(snd_token));
+                }
+            }
+        }
+    }
+
+    for (std::size_t i = 0; i < boost::num_vertices(pattern); ++i) {
+        std::string const& label = pattern[boost::vertex(i, pattern)].attributes.at("label");
+        if (attrs_info.find(label) == attrs_info.end()) {
             continue;
         }
-        HGenerate(current, l, rules, pattern, embeddings, forbidden_rules);
+        for (auto& [name, info] : attrs_info.at(label)) {
+            model::Gfd::Token name_token{i, name};
+            for (auto& value : info) {
+                model::Gfd::Token value_token{-1, value};
+
+                model::Gfd::Literal l = {name_token, value_token};
+                std::set<std::size_t> satisfied_embeddings;
+                std::size_t j = 0;
+                for (Embedding const& embedding : embeddings) {
+                    if (LiteralSatisfied(l, embedding, pattern)) {
+                        satisfied_embeddings.insert(j);
+                    }
+                    j++;
+                }
+                if (satisfied_embeddings.size() >= sigma_) {
+                    satisfied_embeddings_set.try_emplace(l, satisfied_embeddings);
+                    result.emplace_back(name_token, std::move(value_token));
+                }
```

Comment:
That looks very similar with the previous block of code. Maybe extract that to a helper func/lamda to avoid code duplication?